sphinx/ext/autodoc/__init__.py

```diff
@@ -1305,13 +1305,31 @@ def format_args(self, **kwargs: Any) -> str:
     def document_members(self, all_members: bool = False) -> None:
         pass
 
+    def get_canonical_fullname(self) -> Optional[str]:
+        __modname__ = safe_getattr(self.object, '__module__', self.modname)
+        __qualname__ = safe_getattr(self.object, '__qualname__', None)
+        if __qualname__ is None:
+            __qualname__ = safe_getattr(self.object, '__name__', None)
+        if __qualname__ and '<locals>' in __qualname__:
+            # No valid qualname found if the object is defined as locals
+            __qualname__ = None
+
+        if __modname__ and __qualname__:
+            return '.'.join([__modname__, __qualname__])
+        else:
+            return None
+
     def add_directive_header(self, sig: str) -> None:
         sourcename = self.get_sourcename()
         super().add_directive_header(sig)
 
         if inspect.iscoroutinefunction(self.object):
             self.add_line('   :async:', sourcename)
 
+        canonical_fullname = self.get_canonical_fullname()
+        if canonical_fullname and self.fullname != canonical_fullname:
+            self.add_line('   :canonical: %s' % canonical_fullname, sourcename)
```

Comment:
If my understanding is correct, the canonical option is mainly needed for class references because autodoc generates hyperlinks for "real" parent classes and "real" type annotations from its implementation. On the other hand, references to "real" functions will not be generated by autodoc. So I thought canonical-option support for FunctionDocumenter would not be needed.

Reply:
+1

Reply:
Thank you for your wisdom comment!